Update contributing documentation.

cmwshang opened the pull request:
Given the extent of the changes, it may be easier to view the CONTRIBUTING.md and README.md markdown versions in this branch https://github.com/cmwshang/pgbackrest/tree/cs-contributing-ci.

pgstef commented on doc/xml/contributing.xml:
```diff
@@ -118,191 +131,250 @@
+    return THIS_PUB(MyObj)->myData;    // Use the built-in THIS_PUB macro
+}
+
+// TYPE and FROMAT macros for function logging
```
**TYPE and FORMAT** ?

pgstef commented on CONTRIBUTING.md:
```diff
@@ -31,11 +39,171 @@ pgbackrest-dev => Clone pgBackRest repository
+    return THIS_PUB(MyObj)->myData;    // Use the built-in THIS_PUB macro
+}
+
+// TYPE and FROMAT macros for function logging
```
**TYPE and FORMAT** ?

cmwshang replied:
good catch - thanks! Fixed in 269b9f13e715b061cc3d006c62e0b32525af3f3f

pgstef commented on CONTRIBUTING.md:
```diff
@@ -164,189 +331,275 @@ pgbackrest/test/test.pl --vm=u18 --dev --module=mock --test=archive --run=2
+#### Storing a file
 
-To better explain this section, `CFGOPT_ONLINE` will be used as an example:
+Sometimes it is desirable to store or manipulate files before or during a test and then confirm the contents. The [harnessStorage.h](https://github.com/pgbackrest/pgbackrest/blob/master/test/src/common/harnessStorage.h) file contains macros (e.g. `HRN_STORAGE_PUT` and `TEST_STORAGE_GET`) for doing this. In addition, `HRN_INFO_PUT` is convenient for writing out info files (archive.info, backup.inf ...
```
Does this mean that `HRN_STORAGE_PUT` should be used everywhere instead of `TEST_RESULT_VOID(storagePutP(...` or is there some caveats where both should/could be used?

cmwshang replied:
We're trying to encourage everyone to use the newer macros - the old ways still work, but we'd rather not detail the history of the progression to the macros - rather just outline the way it should be done. Maybe we can work on a project to update the tests to be more consistent throughout...

pgstef replied:
Seen the working example in `repo-retention-history` PR. It would perhaps be interesting to update the tests to be consistent but stating that we're now only adding/using new macros is fine :-)

pgstef reviewed (commented):
The `.md` being easier to read, I've left the comments on this rather than on `.xml`.
Not being an English native speaker, I've only found one small typo.

That said, I like this modification. IMHO, the coding example about memory contexts is great. I better understood the layers involved thanks to this example.

The guide says `--coverage-only - selects a vm to use (although best when overridden with --vm=f32)`. I've noticed this too on my laptop. I have to build the f32 vm since it's not built by default.

Shouldn't we change the test script to build the f32 vm by default and use that one for coverage tests then?

dwsteele commented:
> Shouldn't we change the test script to build the f32 vm by default and use that one for coverage tests then?

f32 is obsolete and will be removed soon. We'll be going to u20 for general coverage testing, since the gcc/lcov versions work well together. I think what we need here is to change `--coverage-only` so it does not automatically select a VM.

cmwshang commented:
> > Shouldn't we change the test script to build the f32 vm by default and use that one for coverage tests then?
> 
> f32 is obsolete and will be removed soon. We'll be going to u20 for general coverage testing, since the gcc/lcov versions work well together. I think what we need here is to change `--coverage-only` so it does not automatically select a VM.

Updated in commit 269b9f13e715b061cc3d006c62e0b32525af3f3f : Removed references to --coverage-only since it is now the default for unit tests. Provided more details in the Code Coverage section.

dwsteele commented on doc/xml/contributing.xml:
```diff
@@ -321,87 +393,324 @@
+          option: stanza
+        required: false
+      restore:
+        default: latest
```
The default command role has been renamed to main in PR #1410. So if we commit that first we can add command role here, or leave it in the config.yaml documentation.

pgstef commented on CONTRIBUTING.md:
```diff
@@ -2,7 +2,15 @@
+
+You will always receive credit in the [release notes](http://www.pgbackrest.org/release.html) for your contributions.
+
+Coding standards are defined in [CODING.md](https://github.com/pgbackrest/pgbackrest/blob/master/CODING.md) and some important coding details and an example are provided in the [Coding](#coding) section below. At a minimum, unit tests must be written and run and the documentation generated before submitting a Pull Request; see the [Testing](#testing) section below for details.
```
Perhaps another idea here. Would it be interesting to outline how to work in PR's ? Like "the reviewer will solve the comment once fixed,..."
I know it since you guided me on that path but perhaps it could be nice to write it here too ?

cmwshang replied:
> Perhaps another idea here. Would it be interesting to outline how to work in PR's ? Like "the reviewer will solve the comment once fixed,..."
> I know it since you guided me on that path but perhaps it could be nice to write it here too ?

Agreed, and there is a card to address this in addition to a few other things https://github.com/pgbackrest/pgbackrest/projects/2#card-54174338 - but we are not sure if that will make it into this PR or be worked on in a separate PR.

dwsteele replied:
Agreed, let's keep that for a separate PR.

cmwshang commented on CONTRIBUTING.md:
```diff
@@ -164,189 +329,273 @@ pgbackrest/test/test.pl --vm=u18 --dev --module=mock --test=archive --run=2
-- `TEST_RESULT_UINT` / `TEST_RESULT_INT` - Test for an unsigned integer / integer.
 
-- `TEST_RESULT_BOOL` - Test a boolean return value.
+- `negate` - being a command-line only boolean option, this rule would automatically default to false so it must be defined if the option is negatable. Ask yourself if negation makes sense, for example, would a --dry-run option make sense as --no-dry-run? If the answer is no, then this rule can be omitted as it would automatically default to false. Any boolean option that cannot be negatable, mus ...
```
Fixed, commit bce4ce3e3fb92323d56fd86f66d8caca0fd16df4

cmwshang commented on doc/xml/contributing.xml:
```diff
@@ -118,191 +131,250 @@
-            <title>With Docker</title>
-
-            <p>Build a container to run tests.  The vm must be pre-configured but a variety are available.  The vm names are all three character abbreviations, e.g. <id>u18</id> for <proper>Ubuntu 18.04</proper>.</p>
+        <section id="coding-example">
```
Fixed, commit 4f0913f0db4c68c68f31bc44b01dc8a32d2c9ead

dwsteele reviewed (approved):
This looks good to me now with the updates in c1e933199f2111d304e.
